Eventbrite block: Make widget div ID more unique

Copons opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fix #14560

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Create a unique id for each block instance when the block is rendered.
* Make sure the embedded `iframe` is always as tall as the default/minimum height set by Eventbrite (425px).

Currently the Eventbrite widget uses only the event ID as unique ID, which means that if multiple widgets use the same event, they will conflict between each others.
These conflicts are the cause of several seemingly unrelated issues: block not rendering, blocks rendering inside each other, etc.

| Before | After |
| - | - |
| <img width="492" alt="Screenshot 2020-02-03 at 14 02 09" src="https://user-images.githubusercontent.com/2070010/73659458-33685f80-468e-11ea-92f9-3db5101bd0b2.png"> | <img width="756" alt="Screenshot 2020-02-03 at 17 37 13" src="https://user-images.githubusercontent.com/2070010/73676255-fad77e80-46ab-11ea-8dd3-7421dd1f12b3.png"> |
| <img width="509" alt="Screenshot 2020-02-03 at 14 02 38" src="https://user-images.githubusercontent.com/2070010/73659572-6f9bc000-468e-11ea-8eec-6c2155ac1355.png"> | <img width="752" alt="Screenshot 2020-02-03 at 17 37 40" src="https://user-images.githubusercontent.com/2070010/73676270-03c85000-46ac-11ea-90d5-d2c914eeca6a.png"> |
| <img width="768" alt="Screenshot 2020-02-03 at 14 03 48" src="https://user-images.githubusercontent.com/2070010/73659762-cc977600-468e-11ea-96bb-bdfdb8f512b3.png"> | <img width="752" alt="Screenshot 2020-02-03 at 17 36 52" src="https://user-images.githubusercontent.com/2070010/73676291-0d51b800-46ac-11ea-8146-fa2172856ef1.png"> |

---

The CSS change that forces the container `height: auto !important` and the iframe `height: 425px` is needed to fix an issue occurring when multiple blocks of Embed type use the same event.
The Eventbrite script somehow only applies the correct height (which is both default and minimum 425px) to the last embed, while typically setting the others' height to 0.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Bug fix

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

See the screens in #14560 for more info on what can go wrong.

* Open the editor.
* Insert two Eventbrite widgets using the same event.
* Save and preview the post.
* Make sure both widgets are visible and working as expected.
* Try changing the embed type (both embeds, both buttons, one and one), and make sure the post is always rendered correctly.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A, it's a new block

Copons committed:
[not verified] Add an unique ID to widget divs

Copons committed:
[not verified] Make the iframe height predictable

Copons commented:
@apeatling made an interesting observation: what if Core Gutenberg removes the `clientId`?
The Eventbrite blocks would end up with an incorrect block ID (or worse, might break down altogether).

In 4eb110f, I've added `uuid` to the Jetpack package.json and used it directly instead of indirectly relying on Gutenberg using it for the `clientId`.
I've opted to use the [same version of `uuid` used by Gutenberg (3.3.2)](https://github.com/WordPress/gutenberg/blob/77bd2748461bafcbea7d082aa6824e2473d7890c/package.json#L168), which is already implicitly included in the Jetpack build.

This is the first time I'm touching the Jetpack's package.json, so please be even more careful than usual when reviewing it. 😄

scruffian commented:
Is it likely that Gutenberg would remove the client ID? We also use it in Calendly, Markdown, OpenTable, Repeat Visitor and VideoPress.

apeatling commented:
> Is it likely that Gutenberg would remove the client ID? We also use it in Calendly, Markdown, OpenTable, Repeat Visitor and VideoPress.

Possibly unlikely? Having said that quite a number of fundamentals have changed over the past year, and this is something we'd probably have to be reactive to if it did change.

apeatling commented:
Tested this as per the instructions, both event blocks worked correctly in the editor and front end all scenarios (embed, button, and one and one).

apeatling commented on extensions/blocks/eventbrite/style.scss:
```diff
@@ -1,4 +1,13 @@
+	// Make sure the widget embed container is always as tall as the
+	// default/minimum value set by Eventbrite (425px).
+	&:not( .wp-block-button ) {
+		height: auto !important;
```
You left a good description in the PR of why this !important is needed, I don't think that's communicated in the comment here. It'd be good to be clear so that it doesn't get removed in the future when the context is lost.

Copons replied:
Good point! Updated the comment in 2bb979e04a91d57d6e25114dc724dcbf03edfacd 👍

jeherve reviewed (commented):
Instead of introducing a new dependency, I wonder if we could build the unique ID ourselves, from a timestamp for example? I don't think one would be adding multiple new blocks within the same second for example. 

What do you think?

scruffian commented:
IMO clientID is exactly what this is intended for.

creativecoder reviewed (commented):
I'd also prefer a solution that doesn't involve adding a new js dependency.

Note that if you use the "Duplicate block" feature, you end up with two blocks that have the same `blockId`. I think this attribute should be updated more responsively when `clientId` changes.

creativecoder commented on extensions/blocks/eventbrite/style.scss:
```diff
@@ -1,4 +1,13 @@
+		iframe {
+			height: 425px;
+		}
+	}
```
This seems fragile as Eventbrite could easily change the embed and the 425px height without any notice (but I can't think of a better way to resolve this issue, at the moment) 😕

Copons commented:
> Instead of introducing a new dependency, I wonder if we could build the unique ID ourselves, from a timestamp for example? I don't think one would be adding multiple new blocks within the same second for example.

@jeherve Reason why I've added the dependency is that... it's already added in our build by Gutenberg! 🙂 
Of course, at a second thought, if Gutenberg removes the `clientId`, they would also remove the `uuid` package. 😄 

Using a timestamp, or reverting to `clientId` (especially since it's used in other JP blocks), would be good enough here.

> Note that if you use the "Duplicate block" feature, you end up with two blocks that have the same blockId. I think this attribute should be updated more responsively when clientId changes.

This is a great catch, I didn't think of that. 🤔

Copons committed:
Always update the block ID on block mount, to avoid duplicated ID on block duplication

Copons committed:
Improve inline comment

Copons committed:
Drop unnecessary inline comment

Copons committed:
Use core JS functions instead of the block clientId

Copons commented:
Ok, so in the recent changes I've:
- Reverted the `uuid` implementation.
- Replaced it with a vanilla JS random-ish string.
- And allowed the block to update its `blockId` every time it mounts.

This last change sounds weird, but indeed I don't have a better idea.

Context:
My original idea was to set the `blockId` attribute on block mount _only if the `blockId` wasn't set already_. This way, it was only set at block creation and never again.
As noted by @creativecoder, blocks can also be _duplicated_, and duplicating a block means also copying all its attributes, including the `blockId`.
By preventing the `blockId` overwrite, I was basically letting duplicated blocks have the same `blockId`.
Behind the scene `duplicateBlock` is simply an `insertBlock` with prefilled attributes. As far as I know, there is no way to mark an attribute to be omitted from duplication.
So the only sane solution I've found was to let the block update its `blockId` every time it mounts: which is, AFAIK, every time the editor loads.
Practically this means that the Eventbrite blocks' `blockId` will change every time the post containing them is updated.
I don't think it's a big deal, but let me know if I'm missing something!

Also: I've used this function to create the random ID:
```js
( new Date().getTime() * Math.random() ).toFixed()
```
I did try with just `getTime()` but on my local environment (with ngrok slowing things down) it ended up generating dangerously close timestamps (just 5ms apart for a post with three Eventbrite blocks).

creativecoder commented on extensions/blocks/eventbrite/edit.js:
```diff
@@ -47,8 +47,12 @@ class EventbriteEdit extends Component {
 
+		// Use a random number to make the widget div more unique.
+		setAttributes( { blockId: ( new Date().getTime() * Math.random() ).toFixed() } );
+
```
Updating the `blockId` attribute on component mount does solve the issue when duplicating a block.

However there is now an unfortunate side effect that the editor detects changes simply when loading the page, and displays a confusing warning when trying to navigate away, even without any content changes.

<img width="462" alt="Screen Shot 2020-02-09 at 20 38 30" src="https://user-images.githubusercontent.com/1699996/74117214-8bb2cc00-4b7c-11ea-9b08-b6da26b6f64b.png">

🤔 if there's a better way to handle this... off the top of my head: maybe there's some other prop or attribute  we can check in `componentDidUpdate` that changes when a block is duplicated but not when the block is loaded in the editor? (I'm not sure if clientId works this way, or not)

Copons replied:
I'm kinda tempted to make the block `multiple: false` 😄

scruffian replied:
This is actually a very good idea!

Copons replied:
> This is actually a very good idea!

@scruffian Not really, for at least 2 big reasons! 😄
- Multiple blocks with different events are working fine (and a common enough use case).
- It wouldn't help if the user adds blocks with the same event in different posts that show up in the same archive page.

Copons committed:
Update the block ID if it's a duplicate

jeherve reviewed (dismissed):
This seems to test well for me. It's going to be a bit of a mess for anyone editing this block a few months from now though. I'll merge this for now, but I think we could do with some comments here and there explaining why the block relies on different IDs, and how we handle duplicate blocks.

creativecoder committed:
Move unique id generation for Eventbrite to PHP block rendering

Avoids issues with duplicate ids created within the editor.

creativecoder commented:
After a handoff from @Copons and a smart suggestion from @pablinos , I've refactored so that a unique widget id is generated when the block is rendered (via PHP) rather than in the saved markup. This allows removing a lot of code and not worrying about multiple edge cases around unique ids in the editor.

This needs another round of testing, but generally seems more solid than the other solutions we've been trying. Go team 👯!

pablinos commented on extensions/blocks/eventbrite/save.js:
```diff
@@ -55,7 +49,9 @@ function saveButton( attributes ) {
-				id={ createWidgetId( eventId ) }
+				// Placeholder id, preg replaced with a unique id generated in PHP when the block is rendered.
+				// IMPORTANT: do not remove or change unless you also update the render function in eventbrite.php.
+				id="eventbrite-placeholder-id"
```
I suppose the only problem here is that if the server render function doesn't run you could potentially end up with two elements with the same ID. We could get around that by appending a random number and updating the regex in the PHP. 

It's a super unlikely edge case though, which most browsers would cope with, so it might not be worth bothering with.

creativecoder replied:
That's a great point--I'll make an update to address that.

creativecoder replied:
✅

creativecoder replied:
However, it can't be a random number, because then the block validation fails when the already existing block is loaded in the editor. So I'm just using the `eventId` to help mitigate this.

creativecoder committed:
Fix so that a unique id is saved in the post content

glendaviesnz commented:
Only had time to give it a quick test, but if I added a button then a widget it ended up with the raw eventbrite link between the two on the frontend:

<img width="450" alt="Screen Shot 2020-02-11 at 2 45 49 PM" src="https://user-images.githubusercontent.com/3629020/74205777-946ee380-4cdd-11ea-82ed-a8f0c8ffa90a.png">

creativecoder committed:
Merge branch 'master' into fix/eventbrite-same-event

creativecoder committed:
Remove unused constant

creativecoder committed:
Prevent block validation errors

creativecoder committed:
Fix styles now that markup structure has changed

creativecoder commented:
> but if I added a button then a widget it ended up with the raw eventbrite link between the two on the frontend

Thank you for reviewing! 🤦‍♂ I neglected to update the styles. Should be fixed now.

jeherve reviewed (approved):
I like this, it simplifies things quite nicely! 🚢

jeherve commented:
Cherry-picked to `branch-8.2` in d5870e22e7ea3eea4ced6e820e1891d4a5f681bd

jeherve commented:
r202674-wpcom
